KAFKA-6243: Enable dynamic updates of broker metrics reporters

rajinisivaram opened the pull request:
Dynamic metrics reporter updates described in KIP-226. This includes:
 - Addition and removal of metrics reporters
 - Reconfiguration of custom metrics reporter configs
 - Tests for metrics reporter updates at default cluster-level and as per-broker config for testing

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rajinisivaram committed:
KAFKA-6243: Enable dynamic updates of broker metrics reporters

rajinisivaram commented:
@hachikuji When you have some time, can you review, please? Thank you!

hachikuji commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -218,11 +218,13 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
         reporters.add(new JmxReporter(jmxPrefix))
         val metricConfig = KafkaServer.metricConfig(config)
         metrics = new Metrics(metricConfig, reporters, time, true)
+        brokerMetricsReporters.initMetrics(metrics)
```
This is a little awkward. Couldn't we create the `Metrics` instance with an empty list, then construct the `DynamicMetricsReporters` object and add the reporters?

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -449,3 +456,81 @@ class DynamicLogConfig(logManager: LogManager) extends Reconfigurable with Loggi
+    configs
+  }
+
+  override def validateReconfiguration(configs: util.Map[String, _]): Boolean = {
```
It's a little confusing that validation failure is signaled both by the return value and an exception. Do we need both?

rajinisivaram replied:
As an external API, I thought a boolean is better. For internal validation, I am using exceptions where the message gets propagated to clients (that doesn't feel very safe to do for custom code where exceptions may potentially contain sensitive information like passwords). I have modified `BrokerReconfigurable` to only throw exceptions.

hachikuji replied:
Yes, I think that is fine. Of course plugins may want to propagate a message as well. I guess we could do something like we do with the policies and let the plugin raise `ConfigValidationException` (or something like that).

rajinisivaram replied:
Internally, `ConfigException` is used, which was being translated to `InvalidRequestException` for other config updates (like topic configs). So plugins could throw that exception. I don't mind removing the return value from `validateReconfiguration` if it is better to document the exception instead. Let me know what you think - perhaps I could make that update in the next PR.

hachikuji replied:
Yeah, we can consider separately. I think I slightly prefer documenting the use of `ConfigException` for consistency, but I don't feel too strongly about it.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -449,3 +456,81 @@ class DynamicLogConfig(logManager: LogManager) extends Reconfigurable with Loggi
+    }
+
+    // Validate the new configuration using every reconfigurable reporter instance that is not being deleted
+    currentReporters.values.forall { reporter =>
```
nit: maybe a little more natural to match on the type like above in `reconfigurableConfigs` so we don't need the instance check and cast.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -449,3 +456,81 @@ class DynamicLogConfig(logManager: LogManager) extends Reconfigurable with Loggi
+    }
+  }
+
+  override def reconfigure(configs: util.Map[String, _]): Unit = {
```
Is this any simpler if we use `BrokerReconfigurable`?

rajinisivaram replied:
Since metrics reporters themselves may be `Reconfigurables` with custom configs, it seems simpler to use `Reconfigurable`.

hachikuji commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -642,3 +737,68 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+
+object TestMetricsReporter {
+  val PollingIntervalProp = "polling.interval"
+  val testReporters = new ConcurrentLinkedQueue[TestMetricsReporter]()
```
Should we clear this in a `@Setup`?

hachikuji commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -373,10 +371,101 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+    newProps.put(TestMetricsReporter.PollingIntervalProp, "100")
+    configureMetricsReporters(Seq(classOf[TestMetricsReporter]), newProps)
+
+    def waitForReporters(): List[TestMetricsReporter] = {
```
Could this method go into the `TestMetricsReporter` companion? Maybe the number to wait for could be an argument.

hachikuji commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -373,10 +371,101 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+    newProps.put(TestMetricsReporter.PollingIntervalProp, "4000")
+    alterConfigsOnServer(servers.head, newProps)
+    TestUtils.waitUntilTrue(() => !TestMetricsReporter.testReporters.isEmpty, "Metrics reporter not created")
+    val perBrokerReporter = TestMetricsReporter.testReporters.peek()
```
Is it worth asserting that the size of the collection is 1?

rajinisivaram replied:
I have switched this to use the updated `waitForReporters` method which checks the size.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@hachikuji Thanks for the review, I have made the updates.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -498,39 +492,45 @@ class DynamicMetricsReporters(brokerId: Int, dynamicConfig: DynamicBrokerConfig)
+    currentReporters.values.forall {
+      case reporter: Reconfigurable =>
+        if (updatedMetricsReporters.contains(reporter.getClass.getName))
+          reporter.asInstanceOf[Reconfigurable].validateReconfiguration(configs)
```
We shouldn't need the `asInstanceOf`

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -498,39 +492,45 @@ class DynamicMetricsReporters(brokerId: Int, dynamicConfig: DynamicBrokerConfig)
-        true
+    currentReporters.values.forall {
+      case reporter: Reconfigurable =>
+        if (updatedMetricsReporters.contains(reporter.getClass.getName))
```
Guess you could simplify a little:
```scala
!updatedMetricsReporters.contains(reporter.getClass.getName) || reporter.validateReconfiguration(configs)
```

rajinisivaram committed:
Address review comments

hachikuji commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -218,19 +218,17 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
         val metricConfig = KafkaServer.metricConfig(config)
         metrics = new Metrics(metricConfig, reporters, time, true)
-        brokerMetricsReporters.initMetrics(metrics)
+        new DynamicMetricsReporters(brokerId, config.dynamicConfig, metrics)
```
This looks a little weird because the dynamic config is updated inside the constructor. However, it seems like we could just construct `DynamicMetricsReporters` inside `DynamicBrokerConfig.addReconfigurables` since it already has a reference to the brokerId and we can get `Metrics` from the `KafkaServer` instance.

hachikuji reviewed (commented):
Thanks for the updates. Just a couple more minor comments.

hachikuji reviewed (approved):
LGTM. Thanks for the updates. I will let you merge when ready since jenkins is having issues.

rajinisivaram committed:
Update cluster id for metrics reporters

rajinisivaram commented:
@hachikuji I had missed updating cluster id for metrics reporters that implement `ClusterResourceListener`. I have added that and updated the test. Can you review the last commit, please? Thank you!

hachikuji reviewed (approved):
Good catch. The updates LGTM.

rajinisivaram commented:
Test failures are unrelated, merging to trunk.
